### PR TITLE
feat: per page error handling

### PR DESCRIPTION
**Pull request by ynesteruk**
Improved error handling according to latest changes in `e-commerce-remix` template.

*  Add specific per page error handling with ErrorBoundary

### CODE REVIEW

File: app/routes/products.$categorySlug/route.tsx [hunk 1]
```diff
@@ -113,3 +119,31 @@ export default function ProductsPage() {
...
+export function ErrorBoundary() {
+    const error = useRouteError();
+    const navigate = useNavigate();
+
+    if (isRouteErrorResponse(error)) {
+        let title: string;
+        let message: string | undefined;
+        if (error.data.code === EcomApiErrorCodes.CategoryNotFound) {
+            title = 'Category Not Found';
+            message = "Unfortunately a category you trying to open doesn't exist";
```
Comment (alisey) [inline app/routes/products.$categorySlug/route.tsx:132]: "Unfortunately, the category you're trying to open does not exist."

File: app/routes/thank-you/route.tsx [hunk 2]
```diff
@@ -29,3 +32,9 @@ export default function ThankYouPage() {
         </div>
     );
 }
+
+export function ErrorBoundary() {
+    const error = useRouteError();
+    if (!isRouteErrorResponse(error)) throw error;
+    return <ErrorComponent title="Failed to load product details" message={error.data.message} />;
```
Comment (alisey) [inline app/routes/thank-you/route.tsx:39]: order details

File: src/api/ecom-api.ts [hunk 3]
```diff
@@ -47,100 +41,196 @@ function getWixClient() {
...
-                    .find()
-            ).items;
+                    .hasSome('collectionIds', [category!._id])
+                    .limit(100)
+                    .find();
+
+                const allProducts = productsResponse.items;
+
+                // load all available products. if you have a lot of projects in your site
+                // please implement proper implementation of pagination
```
Comment (alisey) [inline src/api/ecom-api.ts:63]: Remove this crap.

File: app/routes/products.$categorySlug/route.tsx [hunk 4]
```diff
@@ -113,3 +119,31 @@ export default function ProductsPage() {
...
+        } else {
+            title = 'Failed to load category products';
```
Comment (alisey) [inline app/routes/products.$categorySlug/route.tsx:134]: I think a generic "Error" would be better, with the message explaining the details.

File: app/routes/product-details.$productSlug/route.tsx [hunk 5]
```diff
@@ -144,3 +153,31 @@ export default function ProductDetailsPage() {
...
+export function ErrorBoundary() {
+    const error = useRouteError();
+    const navigate = useNavigate();
+
+    if (isRouteErrorResponse(error)) {
+        let title: string;
+        let message: string | undefined;
+        if (error.data.code === EcomApiErrorCodes.ProductNotFound) {
+            title = 'Product Not Found';
+            message = "Unfortunately a product you trying to open doesn't exist";
```
Comment (tetianaman) [inline app/routes/product-details.$productSlug/route.tsx:166]: "product page" to open maybe?

File: app/routes/products.$categorySlug/route.tsx (hunk 1 above)
Comment (tetianaman) [reply app/routes/products.$categorySlug/route.tsx:132]: "category page" maybe ?

File: src/api/ecom-api.ts [hunk 6]
```diff
@@ -17,11 +15,7 @@ function getWixClientId() {
      * or from window.ENV (created by the root loader) on client side.
      */
     const env =
-        typeof window !== 'undefined' && window.ENV
-            ? window.ENV
-            : typeof process !== 'undefined'
-            ? process.env
-            : {};
+        typeof window !== 'undefined' && window.ENV ? window.ENV : typeof process !== 'undefined' ? process.env : {};
```
Comment (tetianaman) [inline src/api/ecom-api.ts:18]: formatting? weird 🤔

File: src/api/ecom-api.ts [hunk 7]
```diff
@@ -47,100 +41,196 @@ function getWixClient() {
...
-        getPromotedProducts: async () => {
-            return (await wixClient.products.queryProducts().limit(4).find()).items;
+
+        async getPromotedProducts() {
+            try {
+                const products = (await wixClient.products.queryProducts().limit(4).find()).items;
+                return successResponse(products);
+            } catch (e) {
+                return failureResponse(EcomApiErrorCodes.GetProductsFailure, getErrorMessage(e));
+            }
```
Comment (tetianaman) [inline src/api/ecom-api.ts:81]: We won't use this function, you can remove it.

File: src/api/ecom-api.ts [hunk 8]
```diff
@@ -47,100 +41,196 @@ function getWixClient() {
...
         },
-        removeItemFromCart: async (id: string) => {
-            const result = await wixClient.currentCart.removeLineItemsFromCurrentCart([id]);
-            return result.cart;
+        async removeItemFromCart(id) {
+            try {
+                const result = await wixClient.currentCart.removeLineItemsFromCurrentCart([id]);
+                if (!result.cart) {
+                    throw new Error('Failed to remove cart item');
```
Comment (Letushev) [inline src/api/ecom-api.ts:130]: Why it throws and not returns failureResponse?

File: app/routes/products.$categorySlug/route.tsx [hunk 9]
```diff
@@ -1,43 +1,49 @@
...
-    } catch (error) {
-        // Redirect to "All products" if the requested category doesn't exist.
-        // In case of another error throw it again and let the global error boundary handle it.
-        if ((error as any)?.details?.applicationError?.code === 404) {
-            return redirectToAllProducts();
-        } else {
-            throw error;
-        }
+    const api = getEcomApi();
+    const currentCategoryResponse = await api.getCategoryBySlug(categorySlug);
```
Comment (Letushev) [inline app/routes/products.$categorySlug/route.tsx:28]: I guess we can use Promise.all here

File: src/api/ecom-api.ts (hunk 7 above)
Comment (ynesteruk) [reply src/api/ecom-api.ts:81]: We won't. What about users who want to built something based on our template?

File: src/api/ecom-api.ts (hunk 8 above)
Comment (ynesteruk) [reply src/api/ecom-api.ts:130]: I geuss to handle all errors in a `catch` block

**Review by Letushev — APPROVED** [review]

**Pull request merged**